wp-includes/class-wp-offline-page-excluder.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * WP_Offline_Page_Excluder class.
+ *
+ * @package PWA
+ */
+
+/**
+ * Class handles excluding the Offline Page's from the backend and frontend.
+ */
+class WP_Offline_Page_Excluder {
+
+	/**
+	 * Instance of the Offline Page Manager.
+	 *
+	 * @var WP_Offline_Page
+	 */
+	protected $manager;
+
+	/**
+	 * WP_Offline_Page_Filter constructor.
+	 *
+	 * @param WP_Offline_Page $manager Instance of the manager.
+	 */
+	public function __construct( $manager ) {
+		$this->manager = $manager;
+	}
+
+	/**
+	 * Initializes the instance.
+	 */
+	public function init() {
+		add_filter( 'wp_dropdown_pages', array( $this, 'exclude_from_page_dropdown' ), 10, 2 );
+		add_action( 'parse_query', array( $this, 'exclude_from_query' ) );
+	}
+
+	/**
+	 * Filters the HTML output of a list of pages as a drop down.
+	 *
+	 * @param string $html HTML output for drop down list of pages.
+	 * @param array  $args The parsed arguments array.
+	 * @return string Filtered content for wp_dropdown_pages.
+	 */
+	public function exclude_from_page_dropdown( $html, $args ) {
+		// Bail out if this is the offline page dropdown.
+		if ( WP_Offline_Page::OPTION_NAME === $args['name'] ) {
+			return $html;
+		}
+
+		return preg_replace( '/<option .*? value="' . $this->manager->get_offline_page_id() . '">.*?<\/option>/', '', $html );
+	}
+
+	/**
+	 * Exclude the offline page from the query when doing a search on the frontend or on the backend Menus page.
+	 *
+	 * @param WP_Query $query The WP_Query instance.
+	 */
+	public function exclude_from_query( WP_Query $query ) {
+		if ( $this->is_offline_page_query( $query ) ) {
+			$query->is_404      = true;
+			$query->is_page     = false;
+			$query->is_singular = false;
+			$query->set( 'page_id', 0 );
+		} elseif ( $this->is_okay_to_exclude( $query ) ) {
+			$query->set( 'post__not_in', array( $this->manager->get_offline_page_id() ) );
```

Comment:
The `post__not_in` query var is advised against for performance reasons: https://vip.wordpress.com/documentation/performance-improvements-by-removing-usage-of-post__not_in/

Maybe that doesn't apply here.

Reply:
@westonruter I consider that in my research.  Let's talk through it to see if we agree that this use case does not apply _or_ if there is a better way to comply but also be performant.

## The Concerns with `post__not_in` var

The concerns about the usage of `post__not_in` var include:

>the query cache hitrate will often be a lot lower

True.  But in this use case, we do not want the offline page included the caching as it should not be included or rendered on the frontend.

>make the query slower _if the exclusion list is large_.

Notice that _slower_ occurs _when the list is large_.  In this use case, our list consists of one (1) item, i.e. one page ID.

## The VIP Proposed Solution - Iterate.

>In almost all cases you can gain great speed improvements by requesting more posts and skipping the posts in PHP.

The proposed alternative solution then is to include the offline page in the query and then remove it or skip over it when iterating through the collection of posts.  

In our use case, this solution would require us to provide the means to exclude the offline page from the collection of posts _after_ we get it back from the database.  Let's think about what this means.

1. The offline page would be cached.  
That's not desirable in our use case.

2. We'd need to remove it from the array of posts.
We could hook into one of the posts filters, such as `posts_results`.  However, iterating over the collection to remove one item is not performant either.

## My Thoughts

I think our use case is the perfect use case for `post__not_in`.  Why?

1. PWA is proposed to be merged into core.
2. We need to exclude the offline page ID from all frontend queries and all searches.
3. It's more performant to exclude in the SQL query than to iterate over the collection of posts for removal.
4. We do not want the offline page to ever be in the cache.

IMO I think using `post__not_in` is the best strategy overall for this use case.

@westonruter What do you think?  Are there other ways we could explore for excluding it from the query that do not involve using `post__not_in`?

Reply:
Your reasoning resonates with me! ✅